### ISSUE TITLE
Graceful Shutdown

### ISSUE THREAD

**Issue by seanmonstar** [carllerche/h2#245]
The `GOAWAY` frame can be used for both errors and for closing cleanly.

> The sender of a GOAWAY frame might gracefully shut down a connection by sending a GOAWAY frame, maintaining the connection in an "open" state until all in-progress streams complete.

This issue is about the 3ish ways of a user initiating a shutdown.

1. **Shutdown immediately**: This is what `server::Connection::close_connection` does currently. Once the `GOAWAY` frame is flushed, the rest of the state machine just exits.
2. **Shutdown gracefully**: This would be when a server no longer wants to accept any more streams than the `last_processed_id`, but will allow reading and writing of still open streams before that ID.
3. **Shutdown notice**: The spec mentions that when a peer wants to start shutting down, it should first send a `GOAWAY` of the max stream ID, and then about 1 RTT later, proceed to our number 2. This allows for any streams that were going to be received around the same time the GOAWAY was flushed to still be accepted, and well-behaving peers SHOULDN'T create anymore streams after those. Of course, bad peers will be bad, but a user of this library can always wrap the shutdown future in a timeout.

We need to decide the API and contract exposed by this library. I suggest this:

1. We change `close_connection` to `close_connection(reason: Reason)`. Behavior is the same. Other names are `disconnect`, `close_immediately`, `shutdown`, `goaway`.
2. We add `close_gracefully()`, which sends a `GOAWAY` of the last processed ID, and then the `Connection` future completes after the allowed streams close up.
3. We add `send_graceful_shutdown()`, which just queues a `GOAWAY` of max stream ID, and nothing else. It's up the user to schedule how they'd like to later call `close_gracefully`.

**Comment by carllerche** [comment]
I believe that this relates to #222, where graceful shutdown was added for the client.

**Comment by carllerche** [comment]
Thanks for doing this write up. This all sounds good.

> The spec mentions that when a peer wants to start shutting down, it should first send a `GOAWAY` of the max stream ID

We should ensure that our client correctly does this.

> and then about 1 RTT later, proceed to our number 2.

The easiest way to do this would be to send a `PING` and wait for the reply.

Re API change:

1) I prefer `close_immediately`. This would also let us deprecate `close_connection` now and impl this behavior.

2+3) Is there a reason to split these up instead of doing the max stream ID bit by default?

**Comment by seanmonstar** [comment]
> The easiest way to do this would be to send a PING and wait for the reply.

We could do that! I looked at a bunch of other HTTP2 implementations, and the only one that does this is the internal h2 implementation of grpc-go. And I found an issue in nghttp2 about fixing a bug to support this, mentioning grpc-go's behavior.

We'd want to be careful if we ever allow the user to send pings with custom payloads, that it didn't conflict with the one we were using, and thus accidentally think we'd seen an ACK much sooner than we should have (on slow connections).

>  Is there a reason to split these up instead of doing the max stream ID bit by default?

Because we needed to wait some period of time, and didn't want to add the need for timers in h2. Assuming a user already needs to wrap shutdown in a timeout, it seemed proper to have this other wait also managed by the user.

If we use a `PING`, then sure, we can combine 2 and 3.
